### PR TITLE
improving setup with compiled c-library, improving the loading of c-l…

### PR DESCRIPTION
**Pull request by Deathn0t**
…ibrary in init, adding version check of c-library, adding simple doc

### CODE REVIEW

File: bindings/python/cconfigspace/__init__.py [hunk 1]
```diff
@@ -1,26 +1,68 @@
...
+import warnings
+
+from ctypes.util import find_library
+
+from .__version__ import __version__
+
+
+__ccs_versions___ = [
+  packaging.version.parse("0.0.0"), 
+  packaging.version.parse("0.0.0")
```
Comment (Kerilk) [inline bindings/python/cconfigspace/__init__.py:14]: I've added a VERSION file in the sources that you should be able to get the max value from.

File: bindings/python/cconfigspace/__version__.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+VERSION = (0, 0, 0)
+
+__version__ = ".".join(map(str, VERSION))
```
Comment (Kerilk) [inline bindings/python/cconfigspace/__version__.py]: `__version__` should be obtained from the new VERSION file.

File: bindings/python/cconfigspace/base.py [hunk 3]
```diff
@@ -7,9 +7,15 @@ class ccs_version(ct.Structure):
               ("minor",    ct.c_ushort),
               ("major",    ct.c_ushort)]
 
-  def __str__(self):
+  def __repr__(self):
     return "{}.{}.{}.{}".format(self.major, self.minor, self.patch, self.revision)
```
Comment (Kerilk) [inline bindings/python/cconfigspace/base.py]: Should be switched to the new format.
```suggestion
    return str(f"{self.major}.{self.minor}.{self.patch}-{self.revision}")
```
Also added the more classical "x.y.z-w" naming scheme.

File: bindings/python/cconfigspace/configuration.py [hunk 4]
```diff
@@ -44,5 +57,28 @@ def configuration_space(self):
...
+    else:
+      raise ValueError("index should be str or int")
+
+  def to_list(self):
+    """Returns a list copy of hyperparameter values of the current configuration."""
+    return self.values[:]
+
+  def to_dict(self):
+    """Returns a dict copy of the current configuration."""
+    return {hp.name:hp_value for hp, hp_value in zip(self.hyperparameters, self.values)}
```
Comment (Kerilk) [inline bindings/python/cconfigspace/configuration.py]: Shouldn't this go in binding.py?

File: bindings/python/cconfigspace/configuration.py [hunk 5]
```diff
@@ -44,5 +57,28 @@ def configuration_space(self):
...
+    elif type(idx) is str:
+      return self.value(idx)
+    else:
+      raise ValueError("index should be str or int")
+
+  def to_list(self):
+    """Returns a list copy of hyperparameter values of the current configuration."""
+    return self.values[:]
+
+  def to_dict(self):
```
Comment (Kerilk) [inline bindings/python/cconfigspace/configuration.py]: Already implemented as asdict in binding.py. I'll let you be judge of the more pythonic option.

File: bindings/python/cconfigspace/configuration.py [hunk 6]
```diff
@@ -44,5 +57,28 @@ def configuration_space(self):
...
+    if type(idx) is int:
+      return self.values[idx]
+    elif type(idx) is str:
+      return self.value(idx)
+    else:
+      raise ValueError("index should be str or int")
+
+  def to_list(self):
+    """Returns a list copy of hyperparameter values of the current configuration."""
+    return self.values[:]
```
Comment (Kerilk) [inline bindings/python/cconfigspace/configuration.py]: Why not alias `to_list` and `values` in binding.py?

File: bindings/python/cconfigspace/configuration.py [hunk 7]
```diff
@@ -44,5 +57,28 @@ def configuration_space(self):
...
+    """
     res = ccs_configuration_check(self.handle)
     Error.check(res)
+
+  @property
+  def hyperparameters(self):
+    """Returns the list of hyperparameters of the current configuration."""
+    return self.configuration_space.hyperparameters
+
+  def __getitem__(self, idx):
```
Comment (Kerilk) [inline bindings/python/cconfigspace/configuration.py]: Why not alias `__getitem__` and `value` in binding.py?
If you want to favor the index code path we can think about switching the isinstance conditions around.

File: bindings/python/setup.py [hunk 8]
```diff
@@ -1,13 +1,40 @@
...
-    )
+  name=NAME,
+  version=about["__version__"],
+  author="Brice Videau",
+  author_email="bvideau@anl.gov",
+  python_requires=REQUIRES_PYTHON,
+  url="https://github.com/argonne-lcf/CCS",
+  packages=[NAME],
+  include_package_data=True,
+  package_data={NAME: [f"libcconfigspace.{get_lib_extension()}"]},
```
Comment (Kerilk) [inline bindings/python/setup.py:42]: This should be optional if the binding is distributed in a platform agnostic way. In this case, bundling the sources and optionally building them if the library is not found on the system would make sense.

File: bindings/python/cconfigspace/configuration.py (hunk 5 above)
Comment (Deathn0t) [reply bindings/python/cconfigspace/configuration.py]: yes but I noticed that this code was faster

File: bindings/python/cconfigspace/__version__.py (hunk 2 above)
Comment (Deathn0t) [reply bindings/python/cconfigspace/__version__.py]: so should I copy the `VERSION` file to `python/cconfigspace/VERSION` when doing `python setup.py install` ? And add this version as data of the package.

File: bindings/python/cconfigspace/configuration.py (hunk 4 above)
Comment (Deathn0t) [reply bindings/python/cconfigspace/configuration.py]: yes we can put it there I did not look at the binding correctly!

**Pull request closed without merging**